<a href="https://colab.research.google.com/github/nishithy13/ds2002-fa26/blob/main/notebooks/01-foundations/2026-09-18%20%E2%80%94%20Pandas_Challenge%20%E2%80%94%20Lab.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# DS2002 · Pandas Challenge

**Lab — 2026-09-18 · Fall 2026**  

---

## Lab 04 — Pandas Challenge

Four hundred generated orders. Each question builds toward a demand report you could hand a vendor.

The data is seeded, so everyone's numbers should match. That is deliberate: if your total revenue differs from your neighbor's, one of you has a bug, and the assertions at the end will tell you which.

Every answer needs the number **and** a sentence saying what it means. A cell that prints `4218.5` with no interpretation is half an answer.

In [1]:
import pandas as pd, numpy as np
rng = np.random.default_rng(4)
n = 400
df = pd.DataFrame({
    'vendor_id': rng.choice(['V-01','V-05','V-10','V-18'], n),
    'category': rng.choice(['Food','Merch','RainGear','Drink'], n, p=[.5,.2,.1,.2]),
    'qty': rng.integers(1, 4, n),
    'price': rng.choice([4.5, 6.0, 7.5, 12.0, 24.0], n),
})
df.head()

,vendor_id,category,qty,price
0,V-10,Drink,2,24.0
1,V-18,RainGear,1,12.0
2,V-18,Drink,3,4.5
3,V-10,Food,2,12.0
4,V-18,Drink,3,7.5


### Q1 — Add `revenue`, then report total revenue and total units.

*Expected: 400 rows, and revenue should land between $8,000 and $9,000.*

In [2]:
# TODO
df['revenue'] = df['qty'] * df['price']

print("Total Revenue:", df['revenue'].sum())
print("Total Units:", df['qty'].sum())

Total Revenue: 8520.0
Total Units: 783


### Q2 — Revenue by category, highest to lowest.

Include the share of total as a percentage in the same table.

In [3]:
# TODO
category_revenue = df.groupby('category')['revenue'].sum().sort_values(ascending = False)

category_revenue = pd.DataFrame(category_revenue)
category_revenue['perecent'] = category_revenue['revenue'] / df['revenue'].sum() * 100

category_revenue

,revenue,perecent
category,,
Food,4293.0,50.387324
Merch,1771.5,20.792254
Drink,1554.0,18.239437
RainGear,901.5,10.580986


### Q3 — Which vendor has the highest *average* order revenue?

Report the average alongside the order count for each vendor. A high average on twelve orders is a different claim from a high average on two hundred.

In [4]:
# TODO
vendor_revenue = df.groupby('vendor_id')['revenue'].agg(['mean', 'count'])

vendor_revenue = vendor_revenue.sort_values('mean', ascending = False)

vendor_revenue

,mean,count
vendor_id,,
V-01,22.595745,94
V-18,21.750000,108
V-05,20.580645,93
V-10,20.314286,105


### Q4 — What share of revenue comes from Merch?

Print it as a percentage rounded to one decimal.

In [5]:
# TODO
merch_revenue = df[df['category'] == 'Merch']['revenue'].sum() / df['revenue'].sum() * 100

print(round(merch_revenue, 1))

20.8


### Q5 — Join in the vendor names.

The frame only has `vendor_id`. Merge the lookup below so your report is readable.

**Requirements:** left join, `validate='many_to_one'`, and prove the row count and revenue total did not change. One vendor id in the orders is not in this lookup — find it, and decide what to do about it.

In [6]:
vendor_names = pd.DataFrame({
    'vendor_id': ['V-01', 'V-05', 'V-10'],
    'vendor_name': ['Hoos Burgers', 'Rotunda Tacos', 'Cav Merch North'],
})

# TODO: merge, validate, and report the unmatched vendor

orders_with_names = df.merge(
    vendor_names,
    on = 'vendor_id',
    how = 'left',
    validate = 'many_to_one')

print(len(df))
print(len(orders_with_names))

print(df['revenue'].sum())
print(orders_with_names['revenue'].sum())

print(orders_with_names[orders_with_names['vendor_name'].isna()])

400
400
8520.0
8520.0
    vendor_id  category  qty  price  revenue vendor_name
1        V-18  RainGear    1   12.0     12.0         NaN
2        V-18     Drink    3    4.5     13.5         NaN
4        V-18     Drink    3    7.5     22.5         NaN
5        V-18     Merch    1   24.0     24.0         NaN
6        V-18     Drink    1   24.0     24.0         NaN
..        ...       ...  ...    ...      ...         ...
384      V-18     Merch    1    4.5      4.5         NaN
386      V-18      Food    3    6.0     18.0         NaN
389      V-18      Food    1    4.5      4.5         NaN
395      V-18     Merch    1   12.0     12.0         NaN
398      V-18     Merch    2   24.0     48.0         NaN

[108 rows x 6 columns]


**The unmatched vendor, and what I did about it:** _..._

### Q6 — A pivot table: vendors down the side, categories across the top, revenue in the cells.

Add row and column totals so it reads as a report rather than a grid of numbers.

In [7]:
# TODO
pivot = df.pivot_table(
    values = 'revenue',
    index = 'vendor_id',
    columns = 'category',
    aggfunc = 'sum',
    margins = True)

pivot

category,Drink,Food,Merch,RainGear,All
vendor_id,,,,,
V-01,171.0,1338.0,373.5,241.5,2124.0
V-05,298.5,882.0,489.0,244.5,1914.0
V-10,502.5,1054.5,400.5,175.5,2133.0
V-18,582.0,1018.5,508.5,240.0,2349.0
All,1554.0,4293.0,1771.5,901.5,8520.0


### Q7 — Validate your work

**TODO:** uncomment and make these pass. Assign your results to the named variables as you go.

In [8]:
# assert len(df) == 400
# assert 8000 < df['revenue'].sum() < 9000, df['revenue'].sum()
# assert abs(by_category['revenue'].sum() - df['revenue'].sum()) < 0.01
# assert len(joined) == len(df), 'the vendor merge changed the row count'
print('checks passed.')

checks passed.


### Write-up

**a)** What would you tell these vendors to do differently next game? One paragraph, with at least two numbers from your report in it.

**b)** Which of your seven answers is the least trustworthy, and why? Point at a specific weakness — a small group size, an unmatched vendor, a category that is really two things.

a) I would tell the vendors to focus more on the categories that generate the most revenue, making sure they have enough inventory for those products. Food generated roughly 50.4% of total revenue. Therefore, they should make sure this category is fully stocked for the next game. On the other hand, Merch pulled in about 20.8% of total revenue. With this, vendors should compare the demands for Food versus Merch to see what is really essential. Lastly, vendors should also use the vendor averages to observe how their individual orders are performing.

b) The least trustworth answers is Q5 with the vendor name report. This is because V-18 was missing from the vendor lookup. The left join kept all 400 ordrs, but V-18 didn't get a vendor name. Unless this gets fixed, the conclusions are not fully accurate.